File: backend/src/main/java/com/example/mypixel/model/node/GaussianBlurNode.java
package com.example.mypixel.model.node;

import com.example.mypixel.exception.InvalidNodeParameter;
import com.example.mypixel.model.ParameterType;
import com.example.mypixel.service.FilteringService;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.lang.NonNull;

import java.util.HashSet;
import java.util.Map;


@MyPixelNode("GaussianBlur")
public class GaussianBlurNode extends Node {

    @Autowired
    private FilteringService filteringService;

    @JsonCreator
    public GaussianBlurNode(
            @JsonProperty("id") @NonNull Long id,
            @JsonProperty("type") @NonNull String type,
            @JsonProperty("inputs") Map<String, Object> inputs) {
        super(id, type, inputs);
    }

    @Override
    public Map<String, ParameterType> getInputTypes() {
        return Map.of(
                "files", ParameterType.FILEPATH_ARRAY.required(),
                "sizeX", ParameterType.INT.required(),
                "sizeY", ParameterType.INT.optional(),
                "sigmaX", ParameterType.DOUBLE.optional(),
                "sigmaY", ParameterType.DOUBLE.optional()
        );
    }

    @Override
    public Map<String, Object> getDefaultInputs() {
        return Map.of(
                "files", new HashSet<String>(),
                "sizeX", 3,
                "sizeY", 3,
                "sigmaX", 0.0,
                "sigmaY", 0.0
        );
    }

    @Override
    public Map<String, ParameterType> getOutputTypes() {
        return Map.of("files", ParameterType.FILEPATH_ARRAY);
    }

    @Override
    public Map<String, String> getDisplayInfo() {
        return Map.of(
                "category", "Filtering",
                "description", "Blurs an image using a Gaussian kernel",
                "color", "#FF8A65",
                "icon", "BlurIcon"
        );
    }

    @Override
    public Map<String, Object> exec() {
        HashSet<String> files = (HashSet<String>) inputs.get("files");
        Map<String, Object> outputs;

        int sizeX = (int) inputs.get("sizeX");
        int sizeY = (int) inputs.getOrDefault("sizeY", sizeX);
        double sigmaX = (double) inputs.getOrDefault("sigmaX", 0.0);
        double sigmaY = (double) inputs.getOrDefault("sigmaY", 0.0);

        batchProcessor.processBatches(files,
                filepath -> filteringService.gaussianBlur(filepath, sizeX, sizeY, sigmaX, sigmaY));

        outputs = Map.of("files", files);

        return outputs;
    }

    @Override
    public void validate() {
        int sizeX = (int) inputs.get("sizeX");
        int sizeY = (int) inputs.getOrDefault("sizeY", sizeX);
        double sigmaX = (double) inputs.getOrDefault("sigmaX", 0.0);
        double sigmaY = (double) inputs.getOrDefault("sigmaY", 0.0);

        if (sizeX < 0 || sizeX % 2 == 0) {
            throw new InvalidNodeParameter("SizeX must be positive and odd");
        }
        if (sizeY < 0 || sizeY % 2 == 0) {
            throw new InvalidNodeParameter("SizeY must be positive and odd");
        }
        if (sigmaX < 0) {
            throw new InvalidNodeParameter("SigmaX must be positive");
        }
        if (sigmaY < 0) {
            throw new InvalidNodeParameter("SigmaY must be positive");
        }
    }
}

File: node/nodes/gaussian_blur_node.py
from typing import Dict, Any
from node import Node


class GaussianBlurNode(Node):

    node_type = "GaussianBlur"

    def get_input_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "files": {
                "type": "FILEPATH_ARRAY",
                "required": True,
                "widget": "LABEL",
                "default": set()
            },
            "sizeX": {
                "type": "INT",
                "required": True,
                "widget": "INPUT",
                "default": 3
            },
            "sizeY": {
                "type": "INT",
                "required": False,
                "widget": "INPUT",
                "default": 3
            },
            "sigmaX": {
                "type": "DOUBLE",
                "required": False,
                "widget": "INPUT",
                "default": 0.0
            },
            "sigmaY": {
                "type": "DOUBLE",
                "required": False,
                "widget": "INPUT",
                "default": 0.0
            }
        }

    def get_output_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "files": {
                "type": "FILEPATH_ARRAY",
                "required": True
            }
        }

    def get_display_info(self) -> Dict[str, str]:
        return {
            "category": "Filtering",
            "description": "Blurs an image using a Gaussian kernel",
            "color": "#FF8A65",
            "icon": "BlurIcon"
        }

    def exec(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
        files = inputs.get("files", set())
        sizeX = inputs.get("sizeX")
        sizeY = inputs.get("sizeY", sizeX)
        sigmaX = inputs.get("sigmaX", 0.0)
        sigmaY = inputs.get("sigmaY", 0.0)

        if not isinstance(files, set):
            files = set(files) if isinstance(files, (list, tuple)) else set()


        return {"files": files}

    def validate(self, inputs: Dict[str, Any]) -> None:
        sizeX = inputs.get("sizeX")
        sizeY = inputs.get("sizeY", sizeX)
        sigmaX = inputs.get("sigmaX", 0.0)
        sigmaY = inputs.get("sigmaY", 0.0)

        try:
            sizeX = int(sizeX)
            sizeY = int(sizeY)
            sigmaX = float(sigmaX)
            sigmaY = float(sigmaY)
        except (TypeError, ValueError):
            raise ValueError("Invalid parameter types")

        if sizeX < 0 or sizeX % 2 == 0:
            raise ValueError("SizeX must be positive and odd")
        if sizeY < 0 or sizeY % 2 == 0:
            raise ValueError("SizeY must be positive and odd")
        if sigmaX < 0:
            raise ValueError("SigmaX must be positive")
        if sigmaY < 0:
            raise ValueError("SigmaY must be positive")
